### PR TITLE
DepositTreeSnapshot Beacon Node API

### PR DESCRIPTION
**Pull request by zilm13**
## PR Description
Introduces Beacon Node API for getting a latest finalized Deposit snapshot tree. Not finished. Part of #5432.
I've several doubts on this:
- as we finished migration could we rename data objects directory to something descriptive like I did or I missed more appropriate directory for this?
- we will need ssz serialization of snapshot tree in several other places in next PRs (loading ssz on Teku startup and storage usage), is it a right place for custom serializer that you suggested to make?
## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Comment by zilm13** [comment]
Added https://github.com/apache/incubator-tuweni/pull/424 to fix Tuweni ssz issue

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/DepositProvider.java [hunk 1]
```diff
@@ -198,6 +200,14 @@ protected synchronized List<DepositWithIndex> getAvailableDeposits() {
     return new ArrayList<>(depositNavigableMap.values());
   }
 
+  protected synchronized Optional<DepositTreeSnapshot> getFinalizedDepositTreeSnapshot() {
+    try {
+      return Optional.of(depositMerkleTree.getSnapshot());
+    } catch (IllegalStateException ex) {
+      return Optional.empty();
+    }
```
Comment (ajsutton) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/DepositProvider.java]: Kind of messy to have to catch `IllegalArgumentException` here.  Probably better to just change `DepositMerkleTree.getSnapshot()`to return an `Optional` rather than throwing an exception if it doesn't have a finalized block.

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/Eth1DataProvider.java [hunk 2]
```diff
@@ -64,6 +67,11 @@ public VotingPeriodInfo getVotingPeriodInfo(final StateAndMetaData stateAndMetaD
     return new VotingPeriodInfo(votesRequired, votingSlots, slotsLeft);
   }
 
+  public SafeFuture<Optional<DepositTreeSnapshotData>> getFinalizedDepositTreeSnapshot() {
+    return SafeFuture.supplyAsync(
```
Comment (ajsutton) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/Eth1DataProvider.java]: Why do we need to run this in a separate thread?  Seems like it could just return `Optional<DepositTreeSnapshotData>`?

File: data/serializer/src/main/java/tech/pegasys/teku/api/data/DepositTreeSnapshotData.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+import java.util.Objects;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.ethereum.pow.merkletree.DepositTreeSnapshot;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class DepositTreeSnapshotData {
+  private final List<Bytes32> finalized;
+  private final UInt64 deposits;
+  private final Bytes32 executionBlockHash;
```
Comment (ajsutton) [inline data/serializer/src/main/java/tech/pegasys/teku/api/data/DepositTreeSnapshotData.java]: This seems to be essentially the same as `DepositTreeSnapshot` so I think we're better off just using `DepositTreeSnapshot` everywhere.  It can move to the `ethereal:pow:api` module so it's available and not just an implementation detail of the merkle tree.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetDepositSnapshot.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        eth1DataProvider
+            .getFinalizedDepositTreeSnapshot()
+            .thenApply(
+                maybeDepositTreeSnapshot -> {
+                  if (maybeDepositTreeSnapshot.isEmpty()) {
+                    return AsyncApiResponse.respondNotFound();
+                  } else {
+                    return AsyncApiResponse.respondOk(maybeDepositTreeSnapshot.get());
+                  }
+                }));
```
Comment (ajsutton) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetDepositSnapshot.java]: There's no need to respond async here - we can just use the normal `request.respondOk` or `request.respondError` methods.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetStateValidators.java [hunk 5]
```diff
@@ -48,7 +48,7 @@
 import java.util.Set;
 import tech.pegasys.teku.api.ChainDataProvider;
 import tech.pegasys.teku.api.DataProvider;
-import tech.pegasys.teku.api.migrated.StateValidatorData;
+import tech.pegasys.teku.api.data.StateValidatorData;
```
Comment (ajsutton) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetStateValidators.java]: I think I'd hold off on doing this rename. It would be better as a separate PR anyway but we're not completely done with the migration yet.  We can rename and potentially reorganise when we actually delete the old rest api types.

File: data/serializer/src/main/java/tech/pegasys/teku/api/response/v1/teku/GetDepositTreeSnapshotResponse.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.swagger.v3.oas.annotations.media.ArraySchema;
+import io.swagger.v3.oas.annotations.media.Schema;
+import java.util.List;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class GetDepositTreeSnapshotResponse {
```
Comment (ajsutton) [inline data/serializer/src/main/java/tech/pegasys/teku/api/response/v1/teku/GetDepositTreeSnapshotResponse.java]: This doesn't match what we'll actually return which has a top level `data` field and then this object.

File: data/serializer/src/main/java/tech/pegasys/teku/api/ssz/DepositTreeSnapshotSerializer.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.ssz.SSZ;
+import tech.pegasys.teku.api.data.DepositTreeSnapshotData;
+import tech.pegasys.teku.infrastructure.http.RestApiConstants;
+import tech.pegasys.teku.infrastructure.restapi.openapi.response.OctetStreamResponseContentTypeDefinition;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class DepositTreeSnapshotSerializer implements SszSerializer<DepositTreeSnapshotData> {
```
Comment (ajsutton) [inline data/serializer/src/main/java/tech/pegasys/teku/api/ssz/DepositTreeSnapshotSerializer.java]: hmm, I was originally expecting this to just be a serialiser for the database format (like we have `BlockCheckpointsSerializer`) but we're having to serialise to SSZ in the REST API and presumably at some point need to deserialise SSZ when the user specifies a snapshot on the command line.  Plus from what I can see the `DepositTreeSnapshot` is only ever used as a data type for this snapshot and serialised as either json or ssz.

So I think I've given you some bad advice and we should just make `DepositTreeSnapshot` an SSZ type and then we get the SSZ serialisation/deserialization plus the JsonTypeDefinition for "free" from that. Sorry.

**Review by ajsutton — COMMENTED** [review]
Looks pretty good.  I played around with the changes I've suggested to make sure I wasn't missing something so have put that up at: https://github.com/zilm13/teku/compare/feature/api-deposit-snapshot...ajsutton:teku:deposit-snapshot-changes 

Feel free to take or leave whatever's useful there.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetStateValidators.java (hunk 5 above)
Comment (zilm13) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetStateValidators.java]: Got you

**Comment by zilm13** [comment]
@ajsutton applied all your changes plus minor cleanup. Ready for review

File: data/serializer/build.gradle [hunk 8]
```diff
@@ -3,12 +3,16 @@ dependencies {
...
+  implementation project(':ethereum:pow:merkletree')
   implementation project(':infrastructure:bls')
   implementation project(':infrastructure:bytes')
   implementation project(':infrastructure:jackson')
   implementation project(':infrastructure:async')
+  implementation project(':infrastructure:restapi')
+  implementation project(':infrastructure:http')
   implementation 'org.apache.tuweni:tuweni-bytes'
   implementation 'org.apache.tuweni:tuweni-units'
+  implementation 'org.apache.tuweni:tuweni-ssz'
```
Comment (ajsutton) [inline data/serializer/build.gradle]: Did we still need all these additions?

File: ethereum/pow/merkletree/src/main/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTree.java [hunk 9]
```diff
@@ -94,12 +95,8 @@ public DepositTree getTreeAtDepositIndex(final long lastDepositIndex) {
...
         lastDepositIndex >= finalizedDepositCount,
         "Cannot recreate tree before finalized deposit count");
-    final DepositTree treeForProof;
-    if (finalizedExecutionBlock.isPresent()) {
-      treeForProof = fromSnapshot(getSnapshot());
-    } else {
-      treeForProof = new DepositTree();
-    }
+    final DepositTree treeForProof =
+        getSnapshot().map(DepositTree::fromSnapshot).orElse(new DepositTree());
```
Comment (ajsutton) [inline ethereum/pow/merkletree/src/main/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTree.java]:
```suggestion
        getSnapshot().map(DepositTree::fromSnapshot).orElseGet(DepositTree::get);
```

to avoid always creating a new DepositTree even if the snapshot is available.

File: ethereum/pow/merkletree/src/test/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTreeTest.java [hunk 10]
```diff
@@ -51,8 +53,8 @@ void shouldGenerateSnapshots(final int nonFinalizedDepositCount) throws Exceptio
       if (i >= nonFinalizedDepositCount) {
         final DepositTestCase finalisingTestCase = testCases.get(i - nonFinalizedDepositCount);
         depositTree.finalize(finalisingTestCase.getEth1Data());
-        final DepositTreeSnapshot snapshot = depositTree.getSnapshot();
-        assertThat(snapshot).isEqualTo(finalisingTestCase.getSnapshot());
+        final Optional<DepositTreeSnapshot> snapshotOptional = depositTree.getSnapshot();
+        assertThat(snapshotOptional.get()).isEqualTo(finalisingTestCase.getSnapshot());
```
Comment (ajsutton) [inline ethereum/pow/merkletree/src/test/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTreeTest.java]:
```suggestion
        assertThat(snapshotOptional).contains(finalisingTestCase.getSnapshot());
```

**Review by ajsutton — APPROVED** [review]
LGTM.

File: ethereum/pow/merkletree/src/main/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTree.java (hunk 9 above)
Comment (zilm13) [reply ethereum/pow/merkletree/src/main/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTree.java]: Nice catch!

File: ethereum/pow/merkletree/src/test/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTreeTest.java (hunk 10 above)
Comment (zilm13) [reply ethereum/pow/merkletree/src/test/java/tech/pegasys/teku/ethereum/pow/merkletree/DepositTreeTest.java]: Thank you

File: data/serializer/build.gradle (hunk 8 above)
Comment (zilm13) [reply data/serializer/build.gradle]: Ouch, old traces remains, cleaning up
